### ISSUE TITLE
Toolbars contributed to the coolbar are no longer visible with Eclipse 4.25

### ISSUE THREAD

**Issue by vaughan82** [eclipse-platform/eclipse.platform.ui#1296]
We have an RCP application that we recently upgraded from using Eclipse 4.16 to Eclipse 4.25 and the toolbars that we add to the CoolbarManager after application initialization, are not displayed. We found, in our test application, that if we use the RMB menu to Lock and then unlock the toolbar area, our toolbars are now displayed. Unfortunately, this workaround is not available in our production application. Thinking it might be an issue with the layout being locked, we have verified that when we add our toolbars, the coolbarManager.getLockLayout API returns false. We have also verified that the toolbars themselves are visible, as are all the buttons in each toolbar at the time the coolbar.update call is made.
We have attempted to work around this issue in a number of different ways, but none have worked, including calling setLockLayout(true) followed by a setLockLayout(false). We view this as a regression in Eclipse.  We would like to see a fix for this in future releases.  We probably also need some type of workaround in the event that we need to change the way we are contributing our toolbars.
Note also - toolbars that we contribute in the ApplicationActionBarAdvisor fillCoolBar method are properly displayed.
This is how we are contributing the majority of our toolbars (again, this occurs after the application is fully initialized and displayed) - we have been doing it this way for many years:
Display.getDefault().asyncExec(new Runnable()
{
public void run()
{
ICoolBarManager cool_bar_mgr = getCoolBarManager();
if (cool_bar_mgr != null && !s_already_populated)
{
s_already_populated = true;

                Vector<ToolbarInfoStruct> toolbar_infos_vec = MyApplication.getEclipseToolbarData();
                
                // For each toolbar ...
                for (int i = 0; i < toolbar_infos_vec.size(); i++)
                {
                    ToolbarInfoStruct tb_info_struct = toolbar_infos_vec.elementAt(i);
                    String tb_english_name = tb_info_struct.getEnglishName();
                    MyToolbarManager tb_mgr = new MyToolbarManager(cool_bar_mgr.getStyle(), tb_english_name);
                    
                    cool_bar_mgr.add(new ToolBarContributionItem(tb_mgr, tb_english_name));
                    
                    s_regID_to_tbMgr_map.put(tb_english_name, tb_mgr);
                    s_localizedID_to_regID.put(
                            tb_info_struct.getLocalizedName(), 
                            tb_english_name);
                    
                    // For each button/separator ...                                    
                    for (int j = 0; j < tb_info_struct.getButtons().size(); j++)
                    {
                        ButtonInfoStruct button_info = tb_info_struct.getButtons().elementAt(j);
                        if (isCmdIDSupported(button_info.getCmdID()))
                        {
                            if (button_info.getCmdID() == -1)
                            {
                                tb_mgr.add(new Separator());
                            }
                            else
                            {
                                try 
                                {
                                    BaseToolbarButtonAction tb_action = null;
                                    
                                    // Check if the button we need is already on a different toolbar. 
                                    Object obj_from_map = s_native_cmd_id_to_action_map.get(button_info.getCmdID());
                                    if (obj_from_map instanceof BaseToolbarButtonAction)
                                    {
                                        // We already have an action for this button, so use the one we've already
                                        // created
                                        tb_action = (BaseToolbarButtonAction)obj_from_map;
                                    }
                                    else
                                    {
                                        // Create a brand-new Action for this button
                                        ImageDescriptor descriptor = button_info.getImageDescriptor();          
                                        String eclipse_cmd_id = s_cmd_starter + "cmd_" + button_info.getCmdID();
                                        tb_action = new BaseToolbarButtonAction(tb_mgr, button_info.getTooltip(), button_info.getCmdID(), eclipse_cmd_id, descriptor);
                                    }
                                
                                    tb_mgr.add(tb_action);
                                    s_native_cmd_id_to_action_map.put(button_info.getCmdID(), tb_action);
                                } 
                                catch (Exception e) 
                                {
                                    e.printStackTrace();
                                }
                            }
                        }
                    }
                    
                    tb_mgr.setVisVisibilityRepresentation(false);
                    tb_mgr.setVisible(false);                                   
                }
            }
            MyApplication.authorizeEclipseToolbarRefresh();
        }
    });         

Toolbar visibility = true is set elsewhere and is then immediately followed by an update to the Coolbar:

    Display.getDefault().syncExec(new Runnable() 
    {
        public void run() 
        {
            ICoolBarManager coolbar_manager = getCoolBarManager();
            if (coolbar_manager != null)
            {   
                coolbar_manager.update(true);
            }   
        }
    });

In order to reproduce this issue: 

The problem is reproducible if you contribute toolbars to the coolbar manager after application initialization has completed.

**Comment by vaughan82** [comment]
I forgot to add, that this issue is urgent.  We have a critical escalated problem report from one of our customers, because they are unable to access functionality that is only available through these exposed toolbars.   We have weekly meetings with our customer and they are looking for answers, which we have not been able to give them.  If the problem could be assigned and investigated, it would be much appreciated.  If there is a workaround to the issue, that would be welcomed.

Thank you.

**Comment by laeubi** [comment]
>  If the problem could be assigned and investigated, it would be much appreciated. 

To whom should it be assigned?

> I forgot to add, that this issue is urgent. 

If this is crucial to your business and likes to speed up the development you can [contact me](https://xn--lubisoft-0za.gmbh/en/#opensource) to enter a contract about this specific issue or if you need dedicated enterprise support.

Of course a PR to fix the problem would also be welcome as it seems you have developers at hand that are familiar with Eclipse RCP, see the [contribution guides here](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md) and if Eclipse is an important Technology for you you maybe even want to join the [Eclipse IDE Workingroup](https://www.eclipse.org/membership/#tab-membership) that also regularly funds development in critical areas or [consider a funding](https://www.eclipse.org/sponsor/ide/) to improve Eclipse IDE in general.

**Comment by iloveeclipse** [comment]
A different question: have you tried with latest Eclipse builds? 4.25 is not supported anymore, current final version is 4.29 and we already have 4.30 M3 shipped last week. Please try with 4.30 M3 if that issue could be already resolved.

If not, please provide a self containing Eclipse project that demonstrates the issue. This would increase the probability that someone will look into your problem.

**Comment by vaughan82** [comment]
We have tried 4.29 and the issue was still present.  We will grab 4.30 M3 to see if it is still an issue there (I am guessing it will be).   We will not be able to move forward from 4.25, the only time the product moves Eclipse versions is with every major release, which occurs every 18 or 24 months.  If we can reproduce the issue with 4.30, we will have to construct a test project to show you the issue.   Thanks.

**Comment by vaughan82** [comment]
We were unable to produce a sample application for you that exhibited the problem, even though we used the exact same sequence of toolbar creation, the application worked.  After some investigation, I now understand why - the sample application would not exit correctly.  I will explain what I found.   Looking at the workspace for our Test application (not the sample we were trying to create), I found the .metadata\.plugins\org.eclipse.e4.workbench\workbench.xmi file, which contained references to our toolbars with visible="false".  This file is written on application exit and sometime after Eclipse 4.16 (we moved from 4.16 to 4.25), always contains our toolbar info with visible="false" (we are setting them false as part of clean up of other components we created).  If I removed that file, the Test application works.  This is why we never saw the issue with the sample application, because it never exited properly.  Note also, that using version of Eclipse 4.16 and prior, our toolbars were not persisted to the workbench.xmi file, thus there was no problem, so this is what caused the problem to appear.  The solution to the issue was to remove our toolbars from the CoolBarManager as the application is going down, so that they are not there to be persisted.
